feat(howto): document the Anbox VHAL HIDL integration into a VHAL implementation (#237)
canonical/anbox-cloud-docs: howto/android/custom-vhal.md

Reviewer: Should we add somewhere in that document that the default Anbox VHAL implements this interface? I'm not sure if it's necessary, but I think it could be interesting information for the user. Wdyt?

Reviewer 2: I have added this information as an intro text

Reviewer: Thanks!

  - custom-vhal
```

> # Integrate Anbox VHAL interface

Reviewer 2: ```suggestion
## Integrate Anbox VHAL interface
```

Author: +1. With the integration of the Anbox VHAL interface, the page 'Customize the Anbox VHAL' becomes complete. Originally, the page was more focused on side-loading a VHAL implementation and making it work with the Anbox runtime via an addon.

> To use and integrate the [Anbox VHAL interface](https://github.com/canonical/vendor_canonical_interfaces/tree/main/vehicle) into a VHAL implementation, the implementation must be built against Android 14. Before proceeding, ensure you have initialized the Android 14 manifest repository. If not, please follow the [official documentation](https://source.android.com/docs/setup/download) to set it up.

Reviewer 2: The spellcheck will still flag initialized because the style switch to US English was just announced. Can we continue using initialised until all infrastructure is in place for the switch? Another option is to add to ignore list but that will keep getting longer.

```suggestion
To use and integrate the [Anbox VHAL interface](https://github.com/canonical/vendor_canonical_interfaces/tree/main/vehicle) into a VHAL implementation, the implementation must be built against Android 14. Before proceeding, ensure you have initialised the Android 14 manifest repository. If you do not yet have the manifest repository ready, follow the [official documentation](https://source.android.com/docs/setup/download) to set it up.
```

Author: +1.
FIxed it

1. Add a new remote definition named `github` to `.repo/manifests/manifest.xml` file.

> 
>        <remote name="aosp"
>           fetch=".."
>           review="https://android-review.googlesource.com/" />
> 
>        <remote name="github"
>           fetch="https://github.com/canonical/" />
> 

Reviewer: Nitpick: could we add the language to have syntactic coloration?
````suggestion
```xml
       <remote name="aosp"
          fetch=".."
          review="https://android-review.googlesource.com/" />

       <remote name="github"
          fetch="https://github.com/canonical/" />
```
````

And same for the ones below.

1. Add a new project named `vendor_canonical_interfaces` to the newly added remote and set it to use the `main` branch.

       <project path="vendor/canonical/interfaces"
                name="vendor_canonical_interfaces"
                remote="github"
                revision="main" />

> 1. Sync the project with the remote

Reviewer 2: ```suggestion
1. Sync the project with the remote.
```

Author: Fixed

      $ repo sync vendor_canonical_interfaces

1. In the VHAL implementation, create a VHAL manifest fragment named `vendor.canonical.interfaces.vehicle@1.0.xml`.

       <manifest version="1.0" type="device">
           <hal format="hidl">
               <name>vendor.canonical.interfaces.vehicle</name>
               <transport>hwbinder</transport>
               <fqname>@1.0::IVehicle/default</fqname>
           </hal>
       </manifest>

>    Place the VHAL manifest fragment file in the same folder as the `Android.bp` file that declares the VHAL service, and include the Anbox VHAL manifest fragment in the VHAL service declaration within the `Android.bp` file. Additionally, add the HIDL module as a shared library that the VHAL service links to in the same file.

Reviewer 2: ```suggestion
   Place the VHAL manifest fragment file in the same folder as the `Android.bp` file that declares the VHAL service, and include the Anbox VHAL manifest fragment in the VHAL service declaration within the `Android.bp` file. Additionally, add the HIDL module as a shared library that the VHAL service links to, in the same VHAL manifest fragment file.
```

Author: hmm, no.
This refers to the Android.bp file, not the device manifest fragment file.
By simply removing "in the same file" the sentence still maintains its meaning but flows more smoothly.
How does that sound?

Reviewer 2: Thanks, that's the kind of clarity we need in such technical context. I think it doesn't hurt to mention the `android.bp` file.

           ],
           shared_libs: [
               ...
               "vendor.canonical.interfaces.vehicle@1.0",
           ],
           ...
       }

1. Take the [example](https://github.com/canonical/vendor_canonical_interfaces/tree/main/vehicle/1.0/default) as a reference, which implements the `IVehicle` interface.

        Return<void> VHalService::get(
            const VehiclePropValue& requestedPropValue, IVehicle::get_cb _hidl_cb) {
          uid_t uid = android::IPCThreadState::self()->getCallingUid();
          if (uid != AID_VEHICLE_NETWORK) {
            _hidl_cb(StatusCode::ACCESS_DENIED, kEmptyValue);
            return Void();
          }

          // NOTE: a VHAL implementation must allow access to non-readable vehicle properties.
          return Void();
        }

        Return<StatusCode> VHalService::set(const VehiclePropValue& value) {
          uid_t uid = android::IPCThreadState::self()->getCallingUid();
          if (uid != AID_VEHICLE_NETWORK)
            return StatusCode::ACCESS_DENIED;

          // NOTE: a VHAL implementation must allow modification of non-writable vehicle properties.
          return StatusCode::NOT_AVAILABLE;
        }

>    Please note that each function must implement a security mechanism to restrict access to vehicle properties to the authorised `AID_VEHICLE_NETWORK` process.

Reviewer: I think this should use a firmer language, e.g.:
```
Please note that you must implement your own access control to vehicle properties to ensure secure access...
```

Or something along those lines. Wdyt?

1. Instantiate the VHAL service that implements the interface and register it as a binder service in the VHAL implementation.

        #include <VHalService.h>

        int main(int /* argc */, char* /* argv */[]) {
            ...
            ...
            configureRpcThreadpool(4, true);
            auto vendor_vhal_service = std::make_unique<VHalService>();
            status_t status = vendor_vhal_service->registerAsService();
            if (status != OK) {
                return 1;
            }
            joinRpcThreadpool();
            return 1;
        }

> After implementing the Anbox HIDL interfaces and integrating them into your VHAL implementation, [build](https://source.android.com/docs/setup/build/building) the VHAL module. Then follow the instructions for [customising the Anbox VHAL](https://documentation.ubuntu.com/anbox-cloud/en/latest/howto/android/custom-vhal/) and load it as an addon during the Android runtime. Once registered, the service can be accessed by the Anbox VHAL adapter.

Reviewer 2: This reads like integrating this interface is the first step and then comes customising Anbox VHAL. But this topic has the instructions in reverse (this is the same custom VHAL topic that is being referred circularly).

Would it make sense to have this integration first in the topic and then the steps about customising VHAL?

Author: That's a really good suggestion. I understand where you're coming from. From my perspective, the integration of the Anbox HIDL interface is a step beyond the first section (customizing VHAL, as I’ve been reluctant to call it), and it requires more low-level knowledge and AOSP development, rather than just following instructions and side-loading a third-party VHAL implementation through an addon to make it work with the Anbox runtime. It's more like an advanced feature for average users.

On the other hand, if this chapter is intended as developer-oriented technical documentation, then yes, we should definitely place the section on integrating the Anbox VHAL at the beginning of this page.

For now, the 'customizing Anbox VHAL' section is more about loading a (third-party) VHAL implementation instead of the default one and making it work with the Anbox runtime. Strictly speaking, it doesn't customize the Anbox VHAL. That's something we might want to consider using a better term for.

Reviewer 2: It is more "custom VHAL implementation" than "customizing the Anbox VHAL". So "custom implementation" is something the user will use instead of the default we provide. It does not mean the user is customizing the Anbox VHAL but rather replaces with their own preferred VHAL implementation. I will make sure it reflects the "custom VHAL implementation" idea everywhere in this topic.
